docs/architecture/adr-008.md

```diff
@@ -0,0 +1,191 @@
+# ADR 008: Denom and Contract Mapping Enhancement for Bi-Directional Token Conversion
+
+## Changelog
+* 2022-06-15: first draft
+
+## Context
+
+Right now Cronos module supports native token to CRC20 token conversion only. As more dApp projects deploy their tokens on Cronos EVM, there is a need to support converting EVM tokens deployed on Cronos EVM to native token for use cases such as bridging to Cosmos Chains by IBC or to Ethereum by Gravity Bridge.
+
+This proposal aims to refine the existing denom and smart contract address mapping design and define the interface of EVM Token to native token conversion. This ADR will also enable tokens to transfer freely between different bridges based on token mappings have only single token source. Applications should design the post-processing mechanisms upon receiving the bridge request. 
+
+## Terminologies
+
+### EVM Token
+
+EVM Tokens are tokens that adopt the widely used CRC20/CRC21 token standard. These token are smart contracts deployed on Cronos EVM.
+
+### Local EVM Token
+
+Local EVM Tokens are EVM tokens that are originated rom Cronos EVM, that means the minting of the tokens happen on Cronos.
+
+### Native Token
+
+Native Tokens are tokens on the Cosmos layer of Cronos. They are not directly usable on Cronos EVM but an intermediate token format between bridge protocols and Cronos EVM because most bridge protocols are Cosmos module that interacts with native tokens.
+
+## Decision
+
+### 1. Denom and Contract Mapping Key-Value
+
+#### A. Current Design
+
+Currently, the native token denom to contract address is registered in the Cronos module kvstore and stored as key-value paris: 
+```
+key -> value
+{prefixDenomToExternalContract}{denom} -> {contractAddress}
+{prefixContractToDenom}{contractAddress} -> {denom}
+```
+where `prefixDenomToExternalContract` and `prefixContractToDenom` are constant integers to identify the types of keys in the key value store. They are `1` and `3` respectively.
+
+For example, IBC-ATOM to contract on Cronos mainnet mapping is
+```
+1ibc/C4CFF46FD6DE35CA4CF4CE031E643C8FDC9BA4B99AE598E9B0ED98FE3A2319F9 -> 0xB888d8Dd1733d72681b30c00ee76BDE93ae7aa93"
+```
+Note that the keys and values are stored as slice of bytes in real life. The above example is for illustration purpose.
+
+#### B. Proposal
+In this ADR, we define an enhanced token mapping key-value pairs as follows:
+
+```
+key -> value
+{prefixDenomToExternalContract}{denom} -> {sourceKey};{contract_address}
+{prefixContractToDenom}{contract_address} -> {sourceKey};{denom}
+```
+where `{sourceKey}` is a list of token sources denoted by integer constants:
+```go
+const (
+    sourceKeyCronos = iota + 1  // 1
+    sourceKeyIBC                // 2
+    sourceKeyGravityBridge      // 3
```

Comment:
ibc and gravity bridge type tokens are distinguishable by prefix of denom: `ibc/***` and `gravity0x***`, I suggest we just define a denom name prefix for the local EVM tokens, `"crc20/" + contract address`.

Reply:
Ok, I think's thats a good point. Then we would require all future tokens to define a unique prefix at their native token, which should be manageable.

Reply:
And maybe we don't need the mapping at all, if we enforce the naming convention `crc20/contract address`.